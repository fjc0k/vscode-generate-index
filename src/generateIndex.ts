import globby from 'globby'
import { IndexGenerator } from './IndexGenerator'
import { Merge } from 'vtils'
import { pathExists, readFile, writeFile } from 'fs-extra'

export interface GenerateIndexPayload {
  filePath: string
  replaceFile?: boolean
  onWarning?: (msg: string) => any
}

export async function generateIndex(
  {
    filePath,
    replaceFile = false,
    onWarning,
  }: GenerateIndexPayload = {} as any,
): Promise<string | false> {
  if (!(await pathExists(filePath))) {
    throw new Error(`File not found. <${filePath}>`)
  }
  const fileContent = (await readFile(filePath)).toString()
  let generatedFileContent = fileContent
  const generator = new IndexGenerator({
    filePath: filePath,
    fileContent: fileContent,
    onWarning: msg => {
      onWarning?.(msg)
    },
    onGenerate: ({ code, marker }) => {
      generatedFileContent =
        generatedFileContent.substr(0, marker.start) +
        code +
        generatedFileContent.substr(marker.end, generatedFileContent.length)
    },
  })
  const generateResult = await generator.generate()
  if (
    generateResult !== false &&
    replaceFile &&
    generatedFileContent !== fileContent
  ) {
    await writeFile(filePath, generatedFileContent)
  }
  return generateResult === false ? false : generatedFileContent
}

export interface GenerateManyIndexPayload {
  patterns: string | string[]
  cwd?: string
  replaceFile: boolean
  onWarning?: (filePath: string, msg: string) => any
  onSuccess?: (filePath: string) => any
}

export interface GenerateManyIndexResult {
  filePath: string
  generatedFileContent: string
}

export async function generateManyIndex(
  {
    patterns,
    cwd = process.cwd(),
    replaceFile = false,
    onWarning,
    onSuccess,
  }: GenerateManyIndexPayload = {} as any,
): Promise<GenerateManyIndexResult[]> {
  if (!patterns.length) return []
  const filePaths = await globby(patterns, {
    absolute: true,
    cwd: cwd,
    dot: true,
    onlyFiles: true,
    unique: true,
    gitignore: false,
  })
  const result = await Promise.all(
    filePaths.map<
      Promise<
        Merge<GenerateManyIndexResult, { generatedFileContent: string | false }>
      >
    >(async filePath => {
      const generatedFileContent = await generateIndex({
        filePath: filePath,
        replaceFile: replaceFile,
        onWarning: msg => onWarning?.(filePath, msg),
      })
      if (generatedFileContent !== false) {
        onSuccess?.(filePath)
      }
      return {
        filePath: filePath,
        generatedFileContent: generatedFileContent,
      }
    }),
  )
  return result.filter(item => item.generatedFileContent !== false) as any
}
