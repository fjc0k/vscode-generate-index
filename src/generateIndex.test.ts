import tmp from 'tempy'
import { dedent } from 'vtils'
import { generateIndex } from './generateIndex'
import { join } from 'path'
import { readFileSync, writeFileSync } from 'fs'
import { sync as rmrf } from 'rimraf'

let dir: string

beforeAll(() => {
  dir = tmp.directory()
  writeFileSync(
    join(dir, 'index.ts'),
    dedent`
      // @index('./*', f => \`export * from '\${f.path}'\`)
      // @endindex
    `,
  )
  writeFileSync(join(dir, 'x.ts'), '')
  writeFileSync(join(dir, '1.ts'), '')
  writeFileSync(join(dir, '国家.ts'), '')
  writeFileSync(join(dir, '_8.ts'), '')
})

afterAll(() => {
  rmrf(dir)
})

test('normal', async () => {
  const generatedContent = await generateIndex(join(dir, 'index.ts'))
  expect(generatedContent).toMatchSnapshot('generated content')
  expect(readFileSync(join(dir, 'index.ts')).toString()).toMatchSnapshot(
    'file content',
  )
})

test('replace', async () => {
  const generatedContent = await generateIndex(join(dir, 'index.ts'), true)
  expect(generatedContent).toMatchSnapshot('generated content')
  expect(readFileSync(join(dir, 'index.ts')).toString()).toMatchSnapshot(
    'file content',
  )
})

test('not found', async () => {
  let err: any
  try {
    await generateIndex(join(dir, 'no-file'))
  } catch (e) {
    err = e
  }
  expect(err).toMatchSnapshot()
})
