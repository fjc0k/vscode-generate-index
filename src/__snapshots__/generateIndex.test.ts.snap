// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`many: res[0].generatedFileContent 1`] = `
"// @index('./*', f => \`export * from '\${f.path}'\`)
export * from './_8'
export * from './1'
export * from './x'
export * from './国家'
// @endindex"
`;

exports[`many: successMsgs 1`] = `
Array [
  "index.ts",
]
`;

exports[`many: warningMsgs 1`] = `
Array [
  "1.ts: No @index maker found.",
  "_8.ts: No @index maker found.",
  "x.ts: No @index maker found.",
  "国家.ts: No @index maker found.",
]
`;

exports[`normal: file content 1`] = `
"// @index('./*', f => \`export * from '\${f.path}'\`)
// @endindex"
`;

exports[`normal: generated content 1`] = `
"// @index('./*', f => \`export * from '\${f.path}'\`)
export * from './_8'
export * from './1'
export * from './x'
export * from './国家'
// @endindex"
`;

exports[`replace: file content 1`] = `
"// @index('./*', f => \`export * from '\${f.path}'\`)
export * from './_8'
export * from './1'
export * from './x'
export * from './国家'
// @endindex"
`;

exports[`replace: generated content 1`] = `
"// @index('./*', f => \`export * from '\${f.path}'\`)
export * from './_8'
export * from './1'
export * from './x'
export * from './国家'
// @endindex"
`;
